server/service/src/stock_line/update.rs

```diff
@@ -96,39 +98,70 @@ fn generate(
         on_hold,
     }: UpdateStockLine,
 ) -> Result<(StockLineRow, Option<Vec<LocationMovementRow>>), UpdateStockLineError> {
-    existing.location_id = location_id.clone().or(existing.location_id);
+    let location_movements = if location_id != existing.location_id {
+        Some(generate_location_movement(
+            store_id,
+            connection,
+            existing.clone(),
+            location_id.clone(),
+        )?)
+    } else {
+        None
+    };
+
+    existing.location_id = location_id.or(existing.location_id);
     existing.batch = batch.or(existing.batch);
     existing.cost_price_per_pack = cost_price_per_pack.unwrap_or(existing.cost_price_per_pack);
     existing.sell_price_per_pack = sell_price_per_pack.unwrap_or(existing.sell_price_per_pack);
     existing.expiry_date = expiry_date.or(existing.expiry_date);
     existing.on_hold = on_hold.unwrap_or(existing.on_hold);
 
-    let location_movements = if location_id != existing.location_id {
-        Some(vec![
-            LocationMovementRow {
-                id: uuid(),
-                store_id: store_id.clone(),
-                stock_line_id: existing.id.clone(),
-                location_id: existing.location_id.clone(),
-                enter_datetime: None,
-                exit_datetime: Some(Utc::now().naive_utc()),
-            },
-            LocationMovementRow {
-                id: uuid(),
-                store_id: store_id,
-                stock_line_id: existing.id.clone(),
-                location_id,
-                enter_datetime: Some(Utc::now().naive_utc()),
-                exit_datetime: None,
-            },
-        ])
-    } else {
-        None
-    };
-
     Ok((existing, location_movements))
 }
 
+fn generate_location_movement(
+    store_id: String,
+    connection: &StorageConnection,
+    existing: StockLineRow,
+    location_id: Option<String>,
+) -> Result<Vec<LocationMovementRow>, UpdateStockLineError> {
+    let mut movement: Vec<LocationMovementRow> = Vec::new();
```

Comment:
:nit movements for the name? I prefer plural for collections

Reply:
Stock line only generates one movement at a time but will use plural for the others (:
[3a7aa5b](https://github.com/openmsupply/open-msupply/pull/812/commits/3a7aa5b82bfb8d06ff2384abc57436d2d5198e77)